### ISSUE TITLE
Support for failures in guards

### ISSUE THREAD

**Issue by hauleth** [elixir-lang/elixir#9704]
In #9470 there was `is_struct/1` macro introduced and `is_struct/2` was rejected due to fact that it should fail when 2nd argument is not atom, i.e.

```elixir
def foo(struct) when is_struct(struct, "bar") or true
```

Should **never** match, just like it happen in case of:

```elixir
def foo(list) when hd(list) == 1 or true
```

And we pass `[]` as an argument.

In [ERL-1122](https://bugs.erlang.org/browse/ERL-1122) it was suggested to use atom `:fail` as `def foo(_) when true and :fail` will always fail match (as `and` requires both sides to return either atom `true` or `false`).

The problem is that currently using:

```elixir
def foo(struct, type) when (is_atom(type) or :fail) and :erlang.map_get(:__struct__, struct) == type
```

Returns many errors. The best solution would be to allow `:fail` or similar non-boolean thing to be allowed in guards. This would allow us to generate given `is_struct/2` macro with ease as:

```elixir
defguard is_struct(map, name) when (is_atom(name) or :fail) and :erlang.map_get(:__struct__, map) == name
```

**Comment by ericmj** [comment]
What errors are you referring to?

**Comment by hauleth** [comment]
My bad, not errors, warnings:

```
==> my_app
Compiling 2 files (.ex)
warning: function clause will never match, found incompatibility:

    :fail !~ boolean()

in expression:

    is_map(params) and is_map_key(params, :__struct__) and is_atom(:erlang.map_get(:__struct__, params)) or :fail

where "params" was given the type %{} in:

    # lib/my_app/controllers/agenda.ex:8
    is_map(params) and is_map_key(params, :__struct__) and is_atom(:erlang.map_get(:__struct__, params)) or :fail

Conflict found at
  lib/my_app/foo.ex:8: MyApp.foo/2
```

**Comment by josevalim** [comment]
This will be closed with #9706. Therefore @hauleth, if you want to submit a PR for `is_struct/2`, it would be very welcome! Thank you!
